### PR TITLE
Fix Broken MIR on match branch simplification w/ u8 #75507

### LINKED ISSUE #75507: LLVM_ERROR with -Zmir-opt-level=3: match.rs
Code from ./src/test/codegen/match.rs
```rust
#![crate_type = "lib"]

pub enum E {
    A,
    B,
}

#[no_mangle]
pub fn exhaustive_match(e: E) -> u8 {
    match e {
        E::A => 0,
        E::B => 1,
    }
}

```

There was an LLVM ERROR with
`rustc ./src/test/codegen/match.rs  -Zmir-opt-level=3`

```
Function return type does not match operand type of return inst!
  ret i1 %7
 i8in function exhaustive_match
LLVM ERROR: Broken function found, compilation aborted!
```
### Meta

repo @ 5e3f1b148db5bfa27fee52464ae1f5d34c49d77b

<details><summary>Backtrace</summary>
<p>

```
Function return type does not match operand type of return inst!
  ret i1 %7
 i8in function exhaustive_match
LLVM ERROR: Broken function found, compilation aborted!
```

</p>
</details>

### PR DESCRIPTION
**Pull request by JulianKnodt**
Fixes #75507 yet, and this adds a regression test.

Done by explicitly checking that the type of the const is a bool.

r? @oli-obk cc: @matthiaskrgr

### CODE REVIEW

**Comment by JulianKnodt** [comment]
I checked and this only breaks on `u8`, which might possibly be because `u8` and `bool` have the same representation.

File: src/test/mir-opt/matches_u8.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// EMIT_MIR_FOR_EACH_BIT_WIDTH
```
Comment (oli-obk) [inline src/test/mir-opt/matches_u8.rs:1]: Are our mir-opt tests running on mir-opt-level 3 by default?

Comment (JulianKnodt) [reply src/test/mir-opt/matches_u8.rs:1]: Uh, I'm not sure, but when I ran it initially with this file it failed

Comment (wesleywiser) [reply src/test/mir-opt/matches_u8.rs:1]: I think they run with `mir-opt-level=2` but I can't find it in the source off hand.

Comment (tmiasko) [reply src/test/mir-opt/matches_u8.rs:1]: https://github.com/rust-lang/rust/blob/8e21bd0633b8d970646ee6eb706c9e8acfad19af/src/tools/compiletest/src/runtest.rs#L1941-L1946

**Comment by oli-obk** [comment]
@bors r-

closing in favour of the other PR that's based on this one

**Pull request closed without merging**